chapter/methodology.md

```diff
@@ -0,0 +1,964 @@
+
+
+        {rank=same; decl_x; x;}
+        {rank=same; def_y; d_y; rec_x}
+        {rank=same; def_z; d_z; field_y}
+}
+```
+
+##### Metadata
+
+In [@sec:meta-information] was shown that on the syntax level, metadata "wraps" the annotated value.
+Conversely, NLS encodes metadata as part of the `LinearizationItem` as it is considered to be intrinsically related to a value.
+NLS therefore has to defer handling of the `MetaValue` node until the processing of the associated value in the succeeding call.
+Like record destructors, NLS temporarily stores this metadata in the `Linearizer`'s memory.
+
+Metadata always precedes its value immediately.
+Thus, whenever a node is linearized, NLS checks whether any latent metadata is stored.
+If there is, it moves it to the value's `LinearizationItem`, clearing the temporary storage.
+
+Although metadata is not linearized as is, contracts encoded in the metadata can however refer to locally bound names.
+Considering that only the annotated value is type-checked and therefore passed to NLS, resolving Usages in contracts requires NLS to separately walk the contract expression.
+Therefore, NLS traverses the AST of expressions used as value annotations.
+In order to avoid interference with the main linearization, contracts are linearized using their own `Linearizer`.
+
+
+### Post-Processing
+Once the entire AST has been processed NLS modifies the Linearization to make it suitable as an efficient index to serve various LSP commands.
+
+After the post-processing the resulting linearization
+
+1. allows efficient lookup of elements from file locations
+2. maintains an `id` based lookup
+3. links deeply nested record destructors to the correct definitions
+4. provides all available type information utilizing Nickel's typing backend
+
+#### Sorting
+
+Since the linearization is performed in a preorder traversal, processing already happens in the order elements are defined physically.
+Yet, during the linearization the location might be unstable or unknown for different items.
+Record fields for instance are processed in an arbitrary order rather than the order they are defined.
+Moreover, for nested records and record short notations, symbolic `Record` items are created which cannot be mapped to a physical location and are thus placed at the range `[0..=0]` in the beginning of the file.
+Maintaining constant insertion performance and item-referencing require that the linearization is exclusively appended.
+Each of these cases, break the physical linearity of the linearization.
+
+NLS thus defers reordering of items.
+The language server uses a stable sorting algorithm to sort items by their associated span's starting position.
+This way, nesting of items with the same start location is preserved.
+Since several operations require efficient access to elements by `id`, which after the sorting does not correspond to the items index in the linearization, after sorting NLS creates an index mapping `id`s to list indices.
```

Comment:
```suggestion
Since several operations require efficient access to elements by `id`, which after the sorting does not correspond to the items index in the linearization, after sorting NLS creates an index mapping `id`s to the new actual indices.
```